### PR TITLE
MessageBatch Interface for Kafka's Iterable<MessageAndOffset>

### PR DESCRIPTION
**Pull request by jenniferdai**
* Allows us to plugin iterables of different types for likafka, aka, indexedRecord
* Removed lag metric because likafka doesn't have support to get the last offset

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 1]
```diff
@@ -355,18 +352,17 @@ protected boolean consumeLoop() throws Exception {
...
-    Iterator<MessageAndOffset> msgIterator = messagesAndOffsets.iterator();
-
     int indexedMessageCount = 0;
     int kafkaMessageCount = 0;
     boolean canTakeMore = true;
     GenericRow decodedRow = null;
     GenericRow transformedRow = null;
-    while (!_shouldStop && !endCriteriaReached() && msgIterator.hasNext()) {
+    int index = 0;
+    while (!_shouldStop && !endCriteriaReached() && index < messagesAndOffsets.getMessageCount()) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java]: I would change this to a 'for' loop to protect against adding a 'continue' statement somewhere in the body of the loop.

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaMessageAndOffset.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.core.realtime.impl.kafka;
+
+import com.linkedin.pinot.core.data.GenericRow;
+
+public class PinotKafkaMessageAndOffset {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaMessageAndOffset.java]: Is this class being used?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumerFactory.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.core.realtime.impl.kafka;
+
+public interface PinotKafkaConsumerFactory {
+  PinotKafkaConsumer buildConsumer(String bootstrapNodes, String clientId, String topic, int partition,
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumerFactory.java]: can u please annotate @Nonnull for bootstrapnodes,clientid, topic, etc.

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumer.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Interface that allows us to plugin different Kafka consumers. The default implementation is SimpleConsumerWrapper.
+ */
+public interface PinotKafkaConsumer extends Closeable {
+  int getPartitionCount(String topic, long timeoutMillis);
+
+  MessageBatch fetchMessages(long startOffset, long endOffset, int timeoutMillis)
+        throws java.util.concurrent.TimeoutException;
+
+  long fetchPartitionOffset(String requestedOffset, int timeoutMillis)
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumer.java]: please add javadoc on what values of 'requestedOffset' is accepted. Can it be null?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/MessageBatch.java [hunk 5]
```diff
@@ -16,13 +16,14 @@
...
  */
-public interface KafkaSimpleConsumerFactory {
-  SimpleConsumer buildSimpleConsumer(String host, int port, int soTimeout, int bufferSize, String clientId);
-}
+public interface MessageBatch<T> {
+    int getMessageCount();
+    T getMessageAtIndex(int index);
+    int getMessageOffsetAtIndex(int index);
+    int getMessageLengthAtIndex(int index);
+    long getNextKafkaMessageOffsetAtIndex(int index);
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/MessageBatch.java]: what does 'getNextKafkaMessageOffsetAtIndex()' do? I am guessing it gets the offset of the next message? What if this is the last message in the batch?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/MessageBatch.java [hunk 6]
```diff
@@ -16,13 +16,14 @@
...
- * Factory for Kafka SimpleConsumer instances, so that we can create mock instances of the Kafka SimpleConsumer for
- * unit tests.
+ * Interface wrapping likafka consumer.
+ * @param <T>
  */
-public interface KafkaSimpleConsumerFactory {
-  SimpleConsumer buildSimpleConsumer(String host, int port, int soTimeout, int bufferSize, String clientId);
-}
+public interface MessageBatch<T> {
+    int getMessageCount();
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/MessageBatch.java]: Please add javadocs, specifying the behavior when index is incorrect, etc.

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 7]
```diff
@@ -102,30 +100,18 @@ public TransientConsumerException(Errors error) {
...
-
-  private SimpleConsumerWrapper(KafkaSimpleConsumerFactory simpleConsumerFactory, String bootstrapNodes,
-      String clientId, String topic, int partition, long connectTimeoutMillis) {
-    _simpleConsumerFactory = simpleConsumerFactory;
+  public SimpleConsumerWrapper(String bootstrapNodes, String clientId, String topic, int partition,
+      long connectTimeoutMillis) {
     _clientId = clientId;
     _topic = topic;
     _partition = partition;
+    if (_topic == null && _partition == Integer.MIN_VALUE) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: I don't understand this. If topic is null and partition is MIN_VALUE, then what can the code above do with the consumer? Both controller and server need to (and will) specify a partition and topic, right?

Comment (jenniferdai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: Topic and partition are ignored for metadata-only consumers. I duplicated the logic we had before for the metadata only consumer

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaMessageAndOffset.java (hunk 2 above)
Comment (jenniferdai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaMessageAndOffset.java]: removed

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java (hunk 1 above)
Comment (jenniferdai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java]: done

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/MessageBatch.java (hunk 5 above)
Comment (jenniferdai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/MessageBatch.java]: Yes, I've duplicated our current logic. currentOffset == finalOffset is checked inside endCriteriaReached.

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumer.java (hunk 4 above)
Comment (jenniferdai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumer.java]: done

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumerFactory.java (hunk 3 above)
Comment (jenniferdai) [reply pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/PinotKafkaConsumerFactory.java]: done

**Pull request merged**